csrc/optimization/remove_empty.cpp

```diff
@@ -0,0 +1,469 @@
+//! of active Expr outputs.
+//!
+//! Derived classes should override handle() and make use of replaceTV(),
+//! markDeadAndMaybeRemove(), and allUsesDead(). Note that if replacements are
+//! made using replaceTV(old_tv, new_tv), then neither new_tv or any new
+//! Statements produced in creating it will be traversed by this class.
+class DeadCodeRemover : BackwardVisitor {
+ public:
+  DeadCodeRemover(Fusion* fusion) : BackwardVisitor(false), fusion_(fusion) {}
+
+  void run() {
+    // First we build a set of all live Statements so that we can detect dead
+    // branches.
+    for (auto stmt : StmtSort::getStmts(fusion_, fusion_->outputs())) {
+      markLive(stmt);
+    }
+
+    // Note that StmtSort::getStmts() is also run in traverseTo. In the future,
+    // we could potentially refactor this so that derived classes from
+    // BackwardVisitor can make use of that traversal instead of repeating it.
+    traverseTo(fusion_, fusion_->outputs());
+  }
+
+  Fusion* fusion() const {
+    return fusion_;
+  }
+
+ protected:
+  using BackwardVisitor::handle;
+
+  void handle(Statement* stmt) final {
+    if (!isLive(stmt)) {
+      // We check whether stmt is dead before we dereference it, since it may
+      // have been removed from the Fusion.
+      return;
+    }
+    BackwardVisitor::handle(stmt);
+  }
+
+  void handle(Expr* expr) final {
+    if (allOutputsDead(expr)) {
+      markDead(expr);
+      fusion_->removeExpr(expr);
+    } else {
+      BackwardVisitor::handle(expr);
+    }
+  }
+
+  //! Replaces a TensorView in outputs, and in all uses. If old_tv is a Fusion
+  //! input, we do not replace it. After replacement, unless it is a Fusion
+  //! input, we remove it from the fusion and set the original pointer to zero
+  //! (hence why old_tv is passed by reference).
+  void replaceTV(TensorView*& old_tv, TensorView* new_tv) {
+    if (old_tv->isFusionOutput()) {
+      fusion_->replaceOutput(old_tv, new_tv);
+    }
+    for (auto use : old_tv->uses()) {
+      ir_utils::replaceValInExpr(use, old_tv, new_tv);
+    }
+    markLiveRecursive(new_tv);
+    markDeadAndMaybeRemove(old_tv);
+  }
+
+  //! Guard removeVal with calls to markDead so that we always detect removed
+  //! Vals and Exprs before derefencing them.
+  //!
+  //! Note that we only remove val if all of its definition's outputs are marked
+  //! dead.
+  void markDeadAndMaybeRemove(Val* val) {
+    markDead(val);
+    if (val->definition()) {
+      if (allOutputsDead(val->definition())) {
+        // If all other outputs are dead, it's safe to remove the definition as
+        // well as all its outputs
+        markDead(val->definition());
+        const auto outputs = val->definition()->outputs();
+        for (auto outp : outputs) {
+          fusion_->removeVal(outp);
+        }
+      }
+    } else {
+      TORCH_INTERNAL_ASSERT(
+          !val->isFusionInput(), "Refusing to remove Fusion input");
+      fusion_->removeVal(val);
```

Comment:
I feel a little nervous with a val is removed while the fusion is being traversed. No particular concern, but is this really safe? A common mistake we have had before was modifying a container while iterating it, so that's why `OptOutMutator` has `registerMutation` and actual mutations are done after a traversal.

Reply:
You're right. This is kind of a time bomb. Instead, it makes sense to probably build up a set of Vals to remove and then do all the removals after the traversal.